### PR TITLE
Armorblox plugin

### PR DESCRIPTION
**Pull request by urajat**
### Description

Describe the proposed changes:

  - Armorblox plugin for InsightConnect

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `make validate` and make sure everything passes
2. Run the assessment tool: `icon-plugin run -A -R all -T all`. For single action validation: `icon-plugin run -A -R tests/my_action.json -T tests/my_action.json`
3. Copy (`icon-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

**Comment by urajat** [comment]
cc @SamhithaTatipalli @absanjay2022

File: plugins/armorblox/icon_armorblox/util/api.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            self.incidents_list.extend(response_json)
+            while next_page_token:
+                params["page_token"] = next_page_token
+                response_json, next_page_token, total_count = self.incidents.list(params=params)
+                self.incidents_list.extend(response_json)
+        except Exception as credentials_exp:
+            print('Exception has occurred for incorrect credentials. ' + str(credentials_exp))
+
+    def get_incidents(self, from_date: str = None, to_date: str = None):
+        params = {
```
Comment (cmcnally-r7) [inline plugins/armorblox/icon_armorblox/util/api.py:28]:
```suggestion
    def get_incidents(self, from_date: str = None, to_date: str = None):
        """
        <description>
        
        :param from_date: <description>
        :param to_date: <description>
        
        :return: <description>
        """
        params = {
```

File: plugins/armorblox/unit_test/payloads/get_remediation_action.json [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      {
+         "detection_tag_id":"4",
+         "detection_tag_name":"invoice"
+      }
+   ],
+   "engagements":{
+      "fwd_mail_count":"0",
+      "reply_mail_count":"0"
+   }
+}
```
Comment (cmcnally-r7) [inline plugins/armorblox/unit_test/payloads/get_remediation_action.json:49]:
```suggestion
}

```

**Review by cmcnally-r7 — CHANGES_REQUESTED** [review]

File: plugins/armorblox/.CHECKSUM
Comment (igorski-r7) [inline plugins/armorblox/.CHECKSUM:1]: Kindly, please add new empty line at the end of file

File: plugins/armorblox/icon_armorblox/connection/connection.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+class Connection(insightconnect_plugin_runtime.Connection):
+
+    def __init__(self):
+        super(self.__class__, self).__init__(input=ConnectionSchema())
+        self.api = None
+
+    def connect(self, params):
+        self.logger.info("Connect: Connecting...")
+        self.logger.info(params)
```
Comment (igorski-r7) [inline plugins/armorblox/icon_armorblox/connection/connection.py:16]: We shouldn't be logging or connection parameters 😄 Please remove it

File: plugins/armorblox/icon_armorblox/connection/connection.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        self.api = ArmorbloxAPI(
+            api_key = params.get('api_key').get('secretKey'),
+            tenant_name = params.get('tenant_name'),
+            logger=self.logger,
```
Comment (igorski-r7) [inline plugins/armorblox/icon_armorblox/connection/connection.py:20]:
```suggestion
            api_key = params.get(Input.API_KEY, {}).get('secretKey'),
            tenant_name = params.get(Input.TENANT_NAME),
            logger=self.logger,
```

File: plugins/armorblox/icon_armorblox/connection/connection.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+    def test(self):
+
+        try:
+            result = self.api.test_api()
+        except PluginException:
+            raise ConnectionTestException(
+                cause="Connection Test Failed.", assistance="Please check that your API key is correct."
+            )
+        return result
```
Comment (igorski-r7) [inline plugins/armorblox/icon_armorblox/connection/connection.py:31]:
```suggestion
        try:
            self.api.test_api()
            return {"success": True}
        except PluginException as error:
            raise ConnectionTestException(
                cause=error.cause, assistance=error.assistance, data=error.data
            )
```

File: plugins/armorblox/icon_armorblox/triggers/get_incidents/trigger.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    def __init__(self):
+        super(self.__class__, self).__init__(
+                name='get_incidents',
+                description=Component.DESCRIPTION,
+                input=GetIncidentsInput(),
+                output=GetIncidentsOutput())
+
+    def run(self, params={}):
+        """Run the trigger"""
+        fetch_interval = int(params.get(Input.INTERVAL))
```
Comment (igorski-r7) [inline plugins/armorblox/icon_armorblox/triggers/get_incidents/trigger.py:22]: I suggest adding the default value 
```suggestion
        fetch_interval = params.get(Input.INTERVAL, DEFAULT_INTERVAL_VALUE)
```

File: plugins/armorblox/icon_armorblox/triggers/get_incidents/trigger.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        # First fetch
+        last_fetch_time = (datetime.utcnow() - timedelta(days=ARMORBLOX_INCIDENT_API_TIME_DELTA_IN_DAYS)).strftime(
+            ARMORBLOX_INCIDENT_API_TIME_FORMAT)
+        while True:
+            current_time = datetime.utcnow().replace(second=0).strftime(ARMORBLOX_INCIDENT_API_TIME_FORMAT)
+            events = self.connection.api.get_incidents(from_date=last_fetch_time, to_date=current_time)
+            self.send({Output.INCIDENTS: events})
```
Comment (igorski-r7) [inline plugins/armorblox/icon_armorblox/triggers/get_incidents/trigger.py:29]: Should we send data if there are no events at all? 😄 
```suggestion
            events = self.connection.api.get_incidents(from_date=last_fetch_time, to_date=current_time)
            if events:
                self.send({Output.INCIDENTS: events})
            else:
                self.logger.info(f"No events retrieved. Sleeping for {fetch_interval} seconds....")
```

File: plugins/armorblox/icon_armorblox/util/api.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        self.incidents_list = []
+        try:
+            response_json, next_page_token, total_count = self.incidents.list(params=params)
+            self.incidents_list.extend(response_json)
+            while next_page_token:
+                params["page_token"] = next_page_token
+                response_json, next_page_token, total_count = self.incidents.list(params=params)
+                self.incidents_list.extend(response_json)
+        except Exception as credentials_exp:
+            print('Exception has occurred for incorrect credentials. ' + str(credentials_exp))
```
Comment (igorski-r7) [inline plugins/armorblox/icon_armorblox/util/api.py:25]: I don't think printing is a good practise here. Kindly, please raise `PluginException` instead.

File: plugins/armorblox/plugin.spec.yaml
Comment (igorski-r7) [inline plugins/armorblox/plugin.spec.yaml:1]: Newline at the end of file please 😄

File: plugins/armorblox/unit_test/test_get_remediation_action.py
Comment (igorski-r7) [inline plugins/armorblox/unit_test/test_get_remediation_action.py:1]: Please add newline at the end of file 😄

File: plugins/armorblox/unit_test/test_get_remediation_action.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+    @patch("requests.get", side_effect=Util.mocked_requests)
+    def test_get_remediation_action(self, mock_post):
+        actual = self.action.run(
+                {
+                    Input.INCIDENT_ID: "63431",
+                }
+            )
+        expected = {'remediation_details': 'WILL_AUTO_REMEDIATE'}
+        self.assertEqual(actual, expected)
```
Comment (igorski-r7) [inline plugins/armorblox/unit_test/test_get_remediation_action.py:28]: Could you please add some more tests for catching the `PluginExceptions`?

**Review by igorski-r7 — CHANGES_REQUESTED** [review]
In general looks good. A few minor changes required.

**Comment by dlaverty-r7** [comment]
@urajat / @SamhithaTatipalli , would you mind closing this PR out if the newer PR supercedes it?

**Comment by dlaverty-r7** [comment]
Closing out as superseded by newer PR.

**Pull request closed without merging**